ouroboros-consensus/src/Ouroboros/Consensus/Mempool/API.hs

```diff
@@ -101,15 +146,27 @@ data Mempool m blk idx = Mempool {
       -- Rejected transactions are therefore not necessarily a sign of
       -- malicious behaviour. Indeed, we would expect /most/ transactions that
       -- are reported as invalid by 'addTxs' to be invalid precisely because
-      -- they have already been included. (Distinguishing between these two
+      -- they have already been included. Distinguishing between these two
       -- cases can be done in theory, but it is expensive unless we have an
-      -- index of transaction hashes that have been included on the blockchain.)
-      addTxs :: [(GenTxId blk, GenTx blk)] -> m [(GenTx blk, ApplyTxErr blk)]
+      -- index of transaction hashes that have been included on the blockchain.
+      addTxs      :: [GenTx blk] -> m [(GenTx blk, Maybe (ApplyTxErr blk))]
 
-      -- | Sync the transactions in the mempool with the ledger state of the
-      -- 'ChainDB'. Invalid transactions will be returned along with the
-      -- validation error.
-    , syncState :: STM m [(GenTx blk, ApplyTxErr blk)]
+      -- | Sync the transactions in the mempool with the current ledger state
+      --  of the 'ChainDB'.
+      --
+      -- The transactions that exist within the mempool will be revalidated
+      -- against the current ledger state. Transactions which are found to be
+      -- invalid, with respect to the current ledger state, will be dropped
+      -- from the mempool, whereas valid transactions will remain.
+      --
+      -- We keep this in @m@ instead of @STM m@ to leave open the possibility
+      -- of persistence. Additionally, this makes it possible to trace the
+      -- removal of invalid transactions.
+      --
+      -- n.b. in our current implementation, when one opens a mempool, we
+      -- spawn a thread which performs this action whenever the 'ChainDB' tip
+      -- point changes.
+    , syncState   :: m ()
```

Comment:
I'm interested to see how we will resolve this being in `m` not `STM m` with the need to select txs when creating a block that are consistent wrt the current ledger state.

Reply:
See #777.